### ISSUE TITLE
`boosting` with `multiprocessing` on Windows

### ISSUE THREAD

**Issue by christianbrodbeck** [Eelbrain/Eelbrain#42]
The `boosting` function has been reported to stall or produce all-zero TRFs when run with multiprocessing on Windows (see https://github.com/Eelbrain/Alice/issues/5).

EDIT: there is a pre-release that works on Windows, see [below](https://github.com/christianbrodbeck/Eelbrain/issues/42#issuecomment-1159544763)

---

Old workarounds
- Use a Linux virtual machine.
- If feasible, the model estimations can be run under a different OS (pickled results can then be loaded and analyzed on Windows, too)
- On Windows, turn off multiprocessing with `eelbrain.configure(n_workers=False)`.

Solutions that don't work
- [Windows Subsystem for Linux](https://docs.microsoft.com/en-us/windows/wsl/install)

**Comment by jkc25** [comment]
Hi Christian, I have a proposed solution to this issue! Instead of using multiprocessing `Process` function, we can use `Thread`. I was inspired by a StackOverflow post that led me to this link ([https://pymotw.com/2/threading/](https://pymotw.com/2/threading/)). This works on Windows; however, I'm not sure if this would work on Linux.
```
threads = []
for _ in range(CONFIG['n_workers']):
   # process = mpc.Process(target=boosting_worker, args=args)
   # process.start()
   t = Thread(target=boosting_worker, args=args)
   threads.append(t)
   t.start() 
```

**Comment by christianbrodbeck** [comment]
Hi @jkc25, this could indeed be a simple solution on Windows, although it probably ought to be platform specific, because threading is usually slower than multiprocessing (by how much depends on the code specifics due to Python's "Global Interpreter Lock"). Have you done any profiling?

**Comment by jkc25** [comment]
Hi @christianbrodbeck, no I have not done any profiling yet to see if it is an improvement over just specifying `eelbrain.configure(n_workers=False)`. My CPU sits at 60% usage when boosting is fitting the models, so I don't think there is any improvement (and I don't think there can be any more improvement due to the Global Interpreter Lock, indeed).
**Update:** Actually, I stand corrected. Using the `Thread` module does have a significant performance boost over using `eelbrain.configure(n_workers=False)`. When using the `multiprocessing` module with no workers model fitting takes ~4 minutes with an average of 5 it/s (avg. of 7% CPU usage), while the `Thread` module takes 40 seconds with an average of 30 it/s (avg. 60% CPU usage).

**Comment by christianbrodbeck** [comment]
I reimplemented boosting with `numba`, this should fix the multiprocessing issues on Windows.

The new implementation is on the [numba-boosting](https://github.com/christianbrodbeck/Eelbrain/tree/numba-boosting) branch.

I also made a pre-release in case anyone wants to test it:
```
conda install -c conda-forge/label/eelbrain_dev -c conda-forge "eelbrain = 0.38a0”
```

@jkc25 , I would be interested in your benchmark if you still have the code for comparison.

EDIT: Note that the progress bar is disabled for now

EDIT 06/21: On macOS the old `multiprocessing`/`Cython` implementation actually seems a bit faster than the new `numba` `parallel` implementation, with the following code (`multiprocessing`: 1.37 s ± 29.9 ms vs. `numba`: 2.29 s ± 38.8 ms):
```py
ds = datasets._get_continuous(100000, True)
%timeit boosting('y', ['x1',  'x2'], 0, 1, ds=ds, test=True, partitions=4)
```
